File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        last: dict[str, int] = {}
        left = 0
        answer = 0

        for right, ch in enumerate(s):
            if ch in last and last[ch] >= left:
                left = last[ch] + 1
            
            last[ch] = right
            answer = max(answer, right - left + 1)

        return answer

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> last;
        int left = 0;
        int ans = 0;

        for (int right = 0; right < (int)s.size(); right++) {
            char ch = s[right];

            if (last.count(ch) && last[ch] >= left) {
                left = last[ch] + 1;
            }

            last[ch] = right;
            ans = max(ans, right - left + 1);
        }

        return ans;
    }
};